### DISCUSSION TITLE
How are loaded libraries used?

### DISCUSSION THREAD

**Discussion by matsekberg** [jingoro2112/wrench#13] [Q&A] [answered]
For example messages.
Is it like msg::read("name")?

**Comment by jingoro2112** [comment, marked as answer]
So... yes and no.

Mostly yes, but technically no.

Yes in the sense that I adopted that convention, but no in the sense that it's only a convention and wrench does not have any explicit support for that nomenclature or namespaces as such.

Let me explain... no there is too much , let me sum up:

The '::' convention is used in wrench to denote "global" scope, in order to guarantee no conflict with existing functions I have library calls use the '::' as part of it's name.

But that support is brute force, functions are registered like this:

`wr_registerLibraryFunction( w, "msg::write", wr_mboxWrite ); 

In code, The wrench parser will recognize the '::' and knows to search ONLY global space for the symbol, but that's as far as it's supported. If you registered the library function as:

`wr_registerLibraryFunction( w, "msg_write", wr_mboxWrite ); 

Or some other plausible name, it would work identically.

**Reply by matsekberg** [reply]
Ahh I see, thanks for the answer!

I like wrench and are about to integrate it in my RV electronics I’m building with esp32’s on custom PCBs.
I receive CanBus messages and process/map them and the payload and send the data over mqtt to displays. The transformation of data is handled by a stack based calculator.
But apart from sending transformed data to displays I also need to take some action depending on states. For example to control a cooling box from battery state and temperature.
For this I need to fetch data from the transformation and use it in wrench.
I was thinking on using the messages library because I thought it could be used as a map or dictionary. But it seems to store integers only. So I guess I will have a c++ map that will be available by both ends.

I really like wrench so far, but some more examples on how to use it and integrate it had made wonders. But I will dive in and get my hands dirty and see if I get it running.

Great job there though!!
